File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.HashSet;

class Solution {
    public boolean solution(String[] phone_book) {
        HashSet<String> set = new HashSet<>();
        
        for (String phone : phone_book) {
            set.add(phone);
        }
        
        for (String phone : phone_book) {
            String parts = "";
            char[] literals = phone.toCharArray();
            for (int i = 0; i < literals.length - 1; i++) {
                parts += literals[i];
                if (set.contains(parts)) {
                    return false;
                }
            }
        }
        
        return true;
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    const set = new Set();
    
    phone_book.forEach(phone => {
        let partial = "";
        for (const literal of phone) {
            partial += literal;
            if (partial !== phone) {
                set.add(partial);
            }
        }
    });
    
    for (const phone of phone_book) {
        if (set.has(phone)) {
            return false;
        }
    }
    
    return true;
}